### DISCUSSION TITLE
User Story: Arranger Clip Editing

### DISCUSSION THREAD

**Discussion by swesterfeld** [tim-janik/anklang#42] [Ideas]
Bob wants to create a piece of music with some clips that are used multiple times in an arrangement. By doing so, he can save some time by repeating patterns of multiple notes. Also, he can change the form of the song later on (by copying and moving clips), without having to use the piano roll at all.

The form Bob wants consists of a melody (Clip A), a second melody (Clip B) and a third piece for the end (Clip C). Each of these clips has two bars. He also wants to have a one bar drum pattern (Clip D), which start later on in the arrangement and is repeated all the time. In the end, the arrangement should look like this:

```
|...A...||...B...||...A....||...B....||...C....|
                  |.D.||.D.||.D.||.D.||.D.||.D.|
```

To achieve this, the following steps are necessary (after setting up the tracks with the instruments):

- double click into the arranger to create clip A
- drag size to two bars
- fill clip A with notes using the piano roll
- do the same steps for clip B
- copy clip A to the destination (using drag + Ctrl)
- copy clip B to the destination (using drag + Ctrl)
- use the same operations to fill and copy clip D to the locations

Nice to have:

- assigning colors to the clips
- assigning names to the clips

**Comment by tim-janik** [comment]
Ok, let's break this down (with a bit of llama.cpp help) into a [Job Story](https://jtbd.info/replacing-the-user-story-with-the-job-story-af7cdee10c27). This can serve as guidance to derive implementation issues, feedback appreciated.

First, analyze the a) situation, b) motivation, c) expected outcome.
Second, derive precise user interface actions, howto steps for the user manual and required software facilities.
Each step involves a specific piece from the user story, a generic UIUX task description, an ITEM title to file this under with a component prefix, a snippet for the DOCS, and a requirement for the CODE.

Here is the first part:

        When I  want to create a piece of music with some clips that are used multiple times in an arrangement ,
        I want to  have a way to easily copy and move these clips around ,
        so I can  avoid repetitive work and focus on creativity.

Second, we derive the list of functional requirements to implement this user story:

1. Users need to be able to create clips in the arranger view.
   UI/UX: Allow creating clips in the arranger view by double-clicking.
   TITLE: "Arranger: add support for creating clips"
   DOCS: "To create a new clip, double-click on an empty space in the arranger view."
   CODE: Implement functionality to create clips in the arranger view when double-clicking.

2. Users need to be able to set the size of clips in the arranger view.
   UI/UX: Allow setting the size of clips by dragging their edges.
   TITLE: "Arranger: add support for changing clip sizes"
   DOCS: "To change the size of a clip, select it and drag its edges to the desired length."
   CODE: Implement functionality to adjust the size of clips in the arranger view.

3. Users need to be able to fill clips with notes using the piano roll.
   UI/UX: Allow filling clips with notes by using the piano roll.
   TITLE: "Piano Roll: add support for creating notes within clips"
   DOCS: "To create notes within a clip, double-click on it and use the piano roll to add notes."
   CODE: Implement functionality to edit notes in the piano roll when a clip is selected.

4. Users need to be able to copy and move clips in the arranger view.
   UI/UX: Allow copying and moving clips by using drag-and-drop with Ctrl or Cmd.
   TITLE: "Arranger: add support for copying and moving clips"
   DOCS: "To copy a clip, select it and drag it to the desired location while holding down Ctrl or Cmd. To move a clip, simply drag it without any additional keys held down."
   CODE: Implement functionality to handle clip duplication and movement in the arranger view.

5. Users need to be able to assign colors and names to clips.
   UI/UX: Allow assigning colors and names to clips through a context menu or similar interface.
   TITLE: "Arranger: add support for coloring and naming clips"
   DOCS: "To color or name a clip, right-click on it and choose 'Assign Color' or 'Set Name' from the context menu."
   CODE: Implement functionality to manage colors and names of clips in the arranger view.

6. Users need to be able to access help and documentation for the arranger.
   UI/UX: Provide accessible help and documentation within the arranger.
   TITLE: "Help: add F1 help for the arranger"
   DOCS: "For help with using the arranger, use the manual in the 'Help' menu or press F1 on your keyboard."
   CODE: Implement F1 handling in the arranger that describes Track and Clip editing.
